### PR TITLE
[design-ish] No false positives for gomod

### PR DESCRIPTION
**Pull request by chmeliik**
# Maintainers will complete the following section

- [ ] Commit messages are descriptive enough
- [ ] Code coverage from testing does not decrease and new code is covered
- [ ] New code has type annotations
- [ ] Docs updated (if applicable)
- [ ] Docs links in the code are still valid (if docs were updated)

### CODE REVIEW

**Comment by chmeliik** [comment]
DO NOT REVIEW THE WHOLE THING

I will try to point out the relevant parts

File: tests/integration/test_gomod.py [hunk 1]
```diff
@@ -252,3 +261,73 @@ def test_e2e_gomod(
...
+    expected_downloaded = {
+        mod
+        for package in output_json["packages"]
+        for dep in package["dependencies"]
+        if dep["type"] == "gomod"
+        if not (mod := _GoModule.from_dict(dep)).is_local()
+    }
+    actual_downloaded = _parse_gomodcache(output_dir / "deps/gomod/pkg/mod")
+
+    assert sorted(map(str, expected_downloaded)) == sorted(map(str, actual_downloaded))
```
[the remark this answers is not in the document]
Comment (chmeliik) [reply tests/integration/test_gomod.py:298]: The plan: get rid of false positives now, enhance inttests with this check after SBOM is implemented.

**Comment by chmeliik** [comment]
PTAL at this "design doc"

TLDR:

When vendoring, we keep the current algorithm
When not vendoring, we use `go mod download -json` and add locally replaced modules based on `go list -deps -json`

Refactor the whole thing to use pydantic models for the go Module and Package structs (as reported in JSON)

Hypothetical minor [breaking change](https://github.com/containerbuildsystem/cachi2/pull/99#discussion_r1140353175)

File: tests/integration/test_gomod.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (chmeliik) [reply tests/integration/test_gomod.py:298]: And for Cachito the Old, do both immediately

**Comment by chmeliik** [comment]
:warning: mistakes were made: `go list -deps -json ./...` does not always report all the modules we expect. With `all` instead of `./...`, it does. I'll go figure out what is the difference.

**Comment by chmeliik** [comment]
> warning mistakes were made: `go list -deps -json ./...` does not always report all the modules we expect. With `all` instead of `./...`, it does. I'll go figure out what is the difference.

Results: https://github.com/chmeliik/godeps#conclusion

Even with `all`, it's not 100% reliable. I'll look into it some more, one solution could be to vendor even in cases when the user doesn't want to vendor (and un-vendor afterwards by removing the vendor dir). Another solution could be to keep `go list -m all` after all. Yet another one could be to shrug and rely on `go list -deps all`

If you need the my conclusions explained better, please let me know.

File: cachi2/core/package_managers/gomod.py [hunk 2]
```diff
@@ -307,60 +282,126 @@ def _resolve_gomod(path: Path, request: Request, git_dir_path=None):
...
+            (name, version)
+            for name, version in map(
+                _get_name_and_version,
+                (
+                    package.module
+                    for package in packages
+                    if package.module and not package.module.main
+                ),
+            )
+            if version.startswith(("./", "../"))
```
[the remark this answers is not in the document]
Comment (chmeliik) [reply cachi2/core/package_managers/gomod.py:322]: This part is wrong, there's a few ways to fix it (see the approaches linked above)

**Comment by chmeliik** [comment]
> Even with `all`, it's not 100% reliable. I'll look into it some more, one solution could be to vendor even in cases when the user doesn't want to vendor (and un-vendor afterwards by removing the vendor dir). Another solution could be to keep `go list -m all` after all (but pick out only the local modules). Yet another one could be to shrug and rely on `go list -deps all`

We agreed to "shrug and rely on `go list -deps all`". 

Closing, will open a cleaner PR soon
